components/brave_today/browser/publishers_controller.cc

```diff
@@ -32,6 +37,40 @@ PublishersController::PublishersController(
 
 PublishersController::~PublishersController() = default;
 
+mojom::PublisherPtr PublishersController::GetPublisherForSite(
+    const GURL& site_url) {
+  if (publishers_.empty())
+    return nullptr;
+
+  const auto& site_origin = url::Origin::Create(site_url);
+  for (const auto& kv : publishers_) {
+    const auto& publisher_origin = url::Origin::Create(kv.second->site_url);
+    if (site_origin.IsSameOriginWith(publisher_origin)) {
+      return kv.second->Clone();
+    }
+  }
+
+  return nullptr;
+}
+
+mojom::PublisherPtr PublishersController::GetPublisherForFeed(
+    const GURL& feed_url) {
+  if (publishers_.empty())
+    return nullptr;
+
+  for (const auto& kv : publishers_) {
+    if (kv.second->feed_source == feed_url)
+      return kv.second->Clone();
+  }
+  return nullptr;
+}
+
+mojom::PublisherPtr PublishersController::GetPublisherById(
+    const std::string& publisher_id) {
+  return publishers_.contains(publisher_id) ? publishers_[publisher_id]->Clone()
```

Comment:
it's better to return raw pointers in these functions. `Clone` is not cheap, callers can do `Clone` if they really need to.

also this exact function should do `find()`, right now it does two lookups.

Reply:
Removed this - I didn't seem to be using it anymore. I've converted the other calls to return a pointer and added a warning about the lifecycle of `publishers_`.